### PR TITLE
ocaml-vdom: init at 0.2

### PR DESCRIPTION
**Pull request by bhootd**
###### Description of changes

Elm architecture and (V)DOM for OCaml 

This package contains:

- OCaml bindings to DOM and other client-side Javascript APIs (using [gen_js_api](https://github.com/LexiFi/gen_js_api)).
- An implementation of the [Elm architecture](https://guide.elm-lang.org/architecture/), where the UI is specified as a functional "view" on the current state.

https://github.com/LexiFi/ocaml-vdom

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/ocaml-modules/ocaml-vdom/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{ lib, fetchurl, buildDunePackage, js_of_ocaml-compiler , gen_js_api }:
+
+buildDunePackage rec {
+  pname = "ocaml-vdom";
+  version = "0.2";
+  minimumOCamlVersion = "4.08";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/ocaml-vdom/default.nix:6]:
```suggestion
  minimalOCamlVersion = "4.08";
```

**Review by vbgl — COMMENTED** [review]
It seems that `ojs` should be listed in `propagatedBuildInputs`.

**Comment by jayesh-bhoot** [comment]
> It seems that `ojs` should be listed in `propagatedBuildInputs`.

I am using this derivation successfully in this flake: https://github.com/jayesh-bhoot/pin-unpin-tab/blob/ocaml-vdom/flake.nix

To me, same logic as I described in https://github.com/NixOS/nixpkgs/pull/172194#discussion_r950890355 should apply here too.

**Comment by jayesh-bhoot** [comment]
> It seems that `ojs` should be listed in `propagatedBuildInputs`.

I think I understand your point now. The project that I was using to test these build derivations is littered with explicit dependencies on the "base" libraries - `js_of_ocaml`, `gen_js_api`, etc.

As soon as I tried using this `ocaml-vdom` derivation standalone (https://github.com/jayesh-bhoot/ocaml-vdom-nix-derivation-demo), I got `Error: Library "ojs" not found` and later `Error: Program js_of_ocaml not found in the tree or in PATH`.

Only after I set these two derivations as `propagatedBuildInput` in this ocaml-vdom derivation did the project compile.

**Comment by jayesh-bhoot** [comment]
I am curious to learn how to decide what goes where. I usually base a derivation on the package's opam file. In `ocaml-vdom`'s case, it does't explicitly list `ojs`. Even its `dune` file does not list an explicit dependency on `ojs`.

How do I figure it out?

**Comment by jayesh-bhoot** [comment]
> but I don’t know about js-of-ocaml-compiler.

I got an error for that when I tried to use the `ocaml-vdom` derivation.

Wow. This is a lot of trial and error.

**Comment by anmonteiro** [comment]
Can we call it `vdom` instead? the `ocaml-*` prefix seems a bit pointless.

**Comment by jayesh-bhoot** [comment]
I wanted to do the same. But

1. The library's opam listing (https://opam.ocaml.org/packages/ocaml-vdom/) uses the `ocaml-vdom`. Better to be consistent.
2. While testing the derivation, dune could find the library only with `ocaml-vdom` (https://github.com/jayesh-bhoot/ocaml-vdom-nix-derivation-demo/blob/main/dune). Interestingly, the examples in their own github repo (https://github.com/LexiFi/ocaml-vdom/blob/master/examples/demo/dune) uses just `vdom`. I couldn't figure out why, except that the `dune` files for those examples may be outdated?

**Comment by anmonteiro** [comment]
> * The library's opam listing (https://opam.ocaml.org/packages/ocaml-vdom/) uses the `ocaml-vdom`. Better to be consistent.

OK this is fair, I had missed it.

For your clarification: you could call it `vdom` in nix and `pname = "ocaml-vdom"` so that Dune could find it.

**Comment by anmonteiro** [comment]
> 2\. While testing the derivation, dune could find the library only with `ocaml-vdom` (https://github.com/jayesh-bhoot/ocaml-vdom-nix-derivation-demo/blob/main/dune). Interestingly, the examples in their own github repo (https://github.com/LexiFi/ocaml-vdom/blob/master/examples/demo/dune) uses just `vdom`. I couldn't figure out why, except that the `dune` files for those examples may be outdated?

yes, it looks like they changed the library name https://github.com/LexiFi/ocaml-vdom/commit/948bcae4ac922e3260f1aee956063f1727dc441d

**Comment by jayesh-bhoot** [comment]
> yes, it looks like they changed the library name https://github.com/LexiFi/ocaml-vdom/commit/948bcae4ac922e3260f1aee956063f1727dc441d

I really should develop a habit to dig into others' source codes, instead of keep guessing.

I will file an issue on their repo tomorrow.

**Pull request merged**